### ISSUE TITLE
Add support for sparsebundles and sparseimages in AppleVM utm bundles

### ISSUE THREAD

**Issue by adespoton** [utmapp/UTM#3478]
For QEMU images, we have qcow2 support.  It would be great to also have sparsebundle/sparseimage support for AppleVM mode.

I personally also use sparsebundles in hand-rolled QEMU with the following method:

hdiutil attach -nomount <file.sparsebundle> 
then use -drive file="/dev/<attached device number of image>",format-raw,media=disk with QEMU.

When QEMU quits, I call hdiutil detach <attached device number of image> to unmount.

But it should be fairly trivial to add native sparse support as well, and even image snapshot support.  Emulators such as Basilisk II and SheepShaver have this built-in.  You can view this at https://github.com/cebix/macemu/blob/4671258ba36cac1a19545b5f666916a30cd91581/BasiliskII/src/Unix/disk_sparsebundle.cpp for a sample implementation (it doesn't implement snapshots).

Sparse snapshots are useful because they're outside the VM implementation; you can set up a disk image the way you want, enable snapshot mounting of the image, and any changes get applied to a second file, so you can always revert to the original.

For that matter, QEMU has this feature with QCOW2 images too, and it will work even on qemu-system-ppc-screamer builds like UTM uses that break VM snapshotting.  So while you can't snapshot a PPC-screamer image runtime, you CAN still create a drive snapshot that you can revert to.  I guess this is a second embedded feature request: drive snapshotting for both QEMU and AppleVM.

Of course, APFS ALSO supports snapshotting, so people can technically snapshot from within the VM.  But I'd argue that it's better to control snapshots from the UTM interface than from the macOS terminal.

**Comment by adespoton** [comment]
Added note: snapshotting on macOS uses hdiutil -shadow.  There's a good example in hdiutil -help that shows how to attach a compressed read-only image as read-write with a shadow file, then when the session is done, how to (optionally) merge the changes in the shadow file back into the read-only image (really, it just replaces the original image with a new one with the new data) using hdiutil convert.

**Comment by osy** [comment]
I think I looked into this and the issue is that Apple Virtualization.framework doesn't support it. The only backing format they support is raw images.

**Comment by adespoton** [comment]
Yes; that’s why I pointed out the two methods of presenting the framework with a virtual raw image, either by using /dev/ or by handling the image format in UTM so it presents a binary data stream to the framework.

**Comment by osy** [comment]
Virtualization.framework runs in a separate process and sandboxed apps can't access block devices or mount images. I am doubtful it could work. I think best bet is to wait for Apple to enhance the framework with more features in the future.
